### PR TITLE
Decouple stream bypass from TLS encrypted bypass v6

### PR DESCRIPTION
**Pull request by lukashino**
Following up on https://github.com/OISF/suricata/pull/11886

Redmine ticket: https://redmine.openinfosecfoundation.org/issues/6788

Describe changes:
v6
- rebased 

v5
- rebased
- added upgrade section
- fixed docs - Thanks Juliana
- SV tests should pass now

v4
- rebased
- changed SSH bypass defaults to hopefully be in sync with the previous settings

v3
- added SSH app-layer option `encryption-handling` allowing to choose whether to continue inspection on SSH once it turns encrypted
- added SV tests
- minor docs updates

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2114

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPR1_alerts_cmp.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .app_layer.flow.dcerpc_tcp | 40 | 42 | 105.0% | 

Pipeline 23244

**Comment by lukashino** [comment]
^ This should get a rerun, kinda weird that it didn't bypass dcerpc where that should not be affected by the PR at all - my assumption is that dcerpc_tcp runs unencrypted.

File: rust/src/ssh/ssh.rs [hunk 1]
```diff
@@ -25,11 +25,16 @@ use crate::frames::Frame;
 
 static mut ALPROTO_SSH: AppProto = ALPROTO_UNKNOWN;
 static HASSH_ENABLED: AtomicBool = AtomicBool::new(false);
+static BYPASS_ENABLED: AtomicBool = AtomicBool::new(false);
 
 fn hassh_is_enabled() -> bool {
     HASSH_ENABLED.load(Ordering::Relaxed)
 }
 
+fn enc_bypass_is_enabled() -> bool {
```
Comment (victorjulien) [inline rust/src/ssh/ssh.rs:34]: style: here and beyond, lets replace `enc` with `encrypt` or `encryption` for readability.

File: src/app-layer-ssh.c [hunk 2]
```diff
@@ -103,6 +105,23 @@ void RegisterSSHParsers(void)
         if (RunmodeIsUnittests() || enable_hassh) {
             rs_ssh_enable_hassh();
         }
+
+        bool enc_bypass = SSH_CONFIG_DEFAULT_BYPASS;
+        ConfNode *enc_handle = ConfGetNode("app-layer.protocols.ssh.encryption-handling");
+        if (enc_handle != NULL && enc_handle->val != NULL) {
+            if (strcmp(enc_handle->val, "full") == 0) {
```
Comment (victorjulien) [inline src/app-layer-ssh.c:112]: "full" to me means `APP_LAYER_PARSER_NO_INSPECTION` isn't set either. It just means full processing and inspection of the rest of the flow even if encrypted phase has started

File: suricata.yaml.in [hunk 3]
```diff
@@ -944,6 +944,13 @@ app-layer:
     ssh:
       enabled: yes
       #hassh: yes
+
+      # What to do when the encrypted communications start:
+      # - bypass:  stop processing this flow as much as possible.
+      #            Offload flow bypass to kernel or hardware if possible.
+      # - full:    keep tracking and inspection as normal
```
Comment (victorjulien) [inline suricata.yaml.in:951]: not what the code does, see above

**Review by victorjulien — CHANGES_REQUESTED** [review]
see inline

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 23247

**Comment by lukashino** [comment]
continues in https://github.com/OISF/suricata/pull/12249

**Pull request closed without merging**